File: src/features/list-books.ts
import { BookStatus, type Book } from '../domain/book.js';
import { type BookStore } from '../store/book-store.js';

/** 空集合提示文案（任务卡 §3.2 输出格式契约）。 */
export const EMPTY_LIST_MESSAGE = '暂无图书';

/** 状态展示标签：`available` → 可借、`borrowed` → 已借出。 */
const STATUS_LABELS: Record<BookStatus, string> = {
  [BookStatus.Available]: '可借',
  [BookStatus.Borrowed]: '已借出',
};

/** FP-005 契约的处理器形状：接收子命令参数，返回结果文本。 */
export type ListHandler = (args: readonly string[]) => string;

/** 命令行分发面的最小契约（FP-005 §3.2 的 `register`）。 */
export interface CommandRegistrar {
  register(name: string, handler: ListHandler): unknown;
}

/** 格式化单本图书为「编号 名称 状态」；未知状态原样回退。 */
export function formatBookLine(book: Book): string {
  const label = STATUS_LABELS[book.status] ?? book.status;
  return `${book.id} ${book.title} ${label}`;
}

/** 逐行格式化全部图书；空集合返回「暂无图书」。 */
export function formatBookList(books: readonly Book[]): string {
  if (books.length === 0) {
    return EMPTY_LIST_MESSAGE;
  }
  return books.map(formatBookLine).join('\n');
}

/**
 * 构造 `list` 处理器：每次调用都读取 `listAll()`，因此状态变化
 * 与集合增删会在下一次查看时即时反映，不做缓存。
 */
export function createListBooksHandler(store: BookStore): ListHandler {
  return () => formatBookList(store.listAll());
}

/** 把 `list` 处理器注册到分发面，返回处理器本身。 */
export function registerListCommand(
  registrar: CommandRegistrar,
  store: BookStore,
): ListHandler {
  const handler = createListBooksHandler(store);
  registrar.register('list', handler);
  return handler;
}

File: library/list_books.py
"""FP-010 图书列表查询与展示：读取全部图书并格式化输出。

只读展示：`list` 子命令的处理器读取数据源的全部图书，每本输出一行
「编号 名称 状态」，状态以「可借 / 已借出」呈现，空集合输出「暂无图书」。
"""

from __future__ import annotations

from typing import Callable, Iterable, List, Protocol

EMPTY_LIST_MESSAGE = "暂无图书"

STATUS_LABELS = {"available": "可借", "borrowed": "已借出"}

Handler = Callable[[List[str]], str]


class BookLike(Protocol):
    """数据契约 FP-001 §3.1 中 Book 的最小结构视图。"""

    id: str
    title: str
    status: str


class BookSource(Protocol):
    """提供全部图书的数据源；FP-001 的 `listAll()` 对应 `list_all()`。"""

    def list_all(self) -> Iterable[BookLike]:
        ...


def status_label(status: str) -> str:
    """把状态值映射为展示文案；未知值原样回退，展示层不因此崩溃。"""
    return STATUS_LABELS.get(status, status)


def format_books(books: Iterable[BookLike]) -> str:
    """逐行格式化图书；空集合返回 `暂无图书`。"""
    lines: List[str] = [
        f"{book.id} {book.title} {status_label(book.status)}" for book in books
    ]
    if not lines:
        return EMPTY_LIST_MESSAGE
    return "\n".join(lines)


def create_list_handler(source: BookSource) -> Handler:
    """构造 `list` 处理器：每次调用都重新读取数据源，不做缓存。"""

    def handler(_args: List[str]) -> str:
        return format_books(source.list_all())

    return handler
